## Part 1: Explore the Data

Import the data and use Pandas to learn more about the dataset.

In [149]:
import pandas as pd

df = pd.read_csv('Resources/client_dataset.csv')
df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number  
0     1096.80     762.71         7.50  105            1  
1       24.95      15.09         1.49   21            0  
2       13.52       7.86         1.68   39            6  
3       36.42      24.85         1.23   29            3  
4      195.10     108.17        46.43   20            1

In [151]:
# View the column names in the data
display(df.columns) # this will print the names of all columns in the dataset


Index(['first', 'last', 'job', 'phone', 'email', 'client_id', 'order_id',
       'order_date', 'order_week', 'order_year', 'item_id', 'category',
       'subcategory', 'unit_price', 'unit_cost', 'unit_weight', 'qty',
       'line_number'],
      dtype='object')

In [153]:
# Use the describe function to gather some basic statistics
display(df.describe()) # this will generate descriptive statistics for numerical columns in the dataset


client_id      order_id    order_week    order_year    unit_price  \
count  54639.000000  5.463900e+04  54639.000000  54639.000000  54639.000000   
mean   54837.869416  5.470190e+06     11.359139   2022.993064    136.267207   
std    25487.438231  2.599807e+06      7.023499      0.082997    183.873135   
min    10033.000000  1.000886e+06      1.000000   2022.000000      0.010000   
25%    33593.000000  3.196372e+06      6.000000   2023.000000     20.800000   
50%    53305.000000  5.496966e+06     11.000000   2023.000000     68.310000   
75%    78498.000000  7.733869e+06     17.000000   2023.000000    173.160000   
max    99984.000000  9.998480e+06     52.000000   2023.000000   1396.230000   

          unit_cost   unit_weight           qty   line_number  
count  54639.000000  54639.000000  5.463900e+04  54639.000000  
mean      99.446073      5.004116  5.702646e+02      2.979667  
std      133.164267      5.326599  1.879552e+04      2.436320  
min        0.010000      0.000000  0.000000e+00      0.000000  
25%       14.840000      1.450000  3.200000e+01      1.000000  
50%       49.890000      3.240000  6.800000e+01      3.000000  
75%      125.570000      6.890000  1.700000e+02      5.000000  
max      846.270000     46.430000  3.958244e+06      9.000000

## Additional Research on Data

In [156]:
# Use this space to do any additional research
# and familiarize yourself with the data.

# Get a summary of the dataset to see the data types and non-null counts
print(df.info()) # this provides a concise summary of the DataFrame, including the data types and non-null counts for each column

# Check for any missing values in the dataset
print(df.isnull().sum()) # this will display the number of missing values in each column

# Display unique values in 'job' column to understand the variety of job titles
print(df['job'].unique()) # this will print all unique job titles in the dataset

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 54639 entries, 0 to 54638
Data columns (total 18 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   first        54639 non-null  object 
 1   last         54639 non-null  object 
 2   job          54639 non-null  object 
 3   phone        54639 non-null  object 
 4   email        54639 non-null  object 
 5   client_id    54639 non-null  int64  
 6   order_id     54639 non-null  int64  
 7   order_date   54639 non-null  object 
 8   order_week   54639 non-null  int64  
 9   order_year   54639 non-null  int64  
 10  item_id      54639 non-null  object 
 11  category     54639 non-null  object 
 12  subcategory  54639 non-null  object 
 13  unit_price   54639 non-null  float64
 14  unit_cost    54639 non-null  float64
 15  unit_weight  54639 non-null  float64
 16  qty          54639 non-null  int64  
 17  line_number  54639 non-null  int64  
dtypes: float64(3), int64(6), object(9)
memory usag

In [158]:
# What three item categories had the most entries?
display(df['category'].value_counts().head(3)) # this will display the three item categories with the most entries

category
consumables    23538
furniture      11915
software        8400
Name: count, dtype: int64

In [160]:
# For the category with the most entries, which subcategory had the most entries?
most_common_category = df['category'].value_counts().idxmax() # find the category with the most entries
display(df[df['category'] == most_common_category]['subcategory'].value_counts().idxmax()) # display the subcategory with the most entries within the most common category


'bathroom supplies'

In [162]:
# Which five clients had the most entries in the data?
display(df['client_id'].value_counts().head(5)) # this will display the five clients with the most entries in the dataset

client_id
33615    220
66037    211
46820    209
38378    207
24741    207
Name: count, dtype: int64

In [164]:
# Store the client ids of those top 5 clients in a list.
top_5_clients = df['client_id'].value_counts().head(5).index.tolist() # store the client ids of the top 5 clients in a list
display(top_5_clients) # this will print the list of the top 5 client ids

[33615, 66037, 46820, 38378, 24741]

In [166]:
# How many total units (the qty column) did the client with the most entries order?
client_with_most_entries = df['client_id'].value_counts().idxmax() # find the client_id with the most entries
total_units_ordered = df[df['client_id'] == client_with_most_entries]['qty'].sum() # sum the qty column for that client_id
display(total_units_ordered) # this will print the total units ordered by the client with the most entries


64313

## Part 2: Transform the Data
Do we know that this client spent the more money than client 66037? If not, how would we find out? Transform the data using the steps below to prepare it for analysis.

In [169]:
# Create a column that calculates the subtotal for each line using the unit_price and the qty
df['subtotal'] = df['unit_price'] * df['qty'] # create a new column 'subtotal' by multiplying 'unit_price' and 'qty'
display(df[['unit_price', 'qty', 'subtotal']].head()) # this will display the first few rows to verify the new column

unit_price  qty   subtotal
0     1096.80  105  115164.00
1       24.95   21     523.95
2       13.52   39     527.28
3       36.42   29    1056.18
4      195.10   20    3902.00

In [171]:
# Create a column for shipping price.
# Assume a shipping price of $7 per pound for orders over 50 pounds and $10 per pound for items 50 pounds or under.
df['shipping_price'] = df['unit_weight'].apply(lambda x: 7 if x > 50 else 10) * df['unit_weight'] # calculate shipping price based on unit_weight
display(df[['unit_weight', 'shipping_price']].head()) # this will display the first few rows to verify the new column

unit_weight  shipping_price
0         7.50            75.0
1         1.49            14.9
2         1.68            16.8
3         1.23            12.3
4        46.43           464.3

In [173]:
# Create a column for the total price using the subtotal and the shipping price along with a sales tax of 9.25%
df['total_price'] = (df['subtotal'] + df['shipping_price']) * 1.0925 # calculate the total price including sales tax
display(df.head()) # this will display the first few rows to verify the new column


first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year  ...     category         subcategory unit_price  \
0          17        2023  ...        decor            wall art    1096.80   
1          20        2023  ...  consumables                pens      24.95   
2           4        2023  ...     software  project management      13.52   
3          17        2023  ...  consumables                pens      36.42   
4          21        2023  ...  consumables                misc     195.10   

   unit_cost  unit_weight  qty  line_number   subtotal  shipping_price  \
0     762.71         7.50  105            1  115164.00            75.0   
1      15.09         1.49   21            0     523.95            14.9   
2       7.86         1.68   39            6     527.28            16.8   
3      24.85         1.23   29            3    1056.18            12.3   
4     108.17        46.43   20            1    3902.00           464.3   

     total_price  
0  125898.607500  
1     588.693625  
2     594.407400  
3    1167.314400  
4    4770.182750  

[5 rows x 21 columns]

In [175]:
# Create a column for the cost of each line using unit cost, qty, and shipping price
# (assume the shipping cost is exactly what is charged to the client)
df['line_cost'] = (df['unit_cost'] * df['qty']) + df['shipping_price'] # calculate the total cost for each line
display(df.head()) # this will display the first few rows to verify the new column




first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year  ...         subcategory unit_price unit_cost  \
0          17        2023  ...            wall art    1096.80    762.71   
1          20        2023  ...                pens      24.95     15.09   
2           4        2023  ...  project management      13.52      7.86   
3          17        2023  ...                pens      36.42     24.85   
4          21        2023  ...                misc     195.10    108.17   

   unit_weight  qty  line_number   subtotal  shipping_price    total_price  \
0         7.50  105            1  115164.00            75.0  125898.607500   
1         1.49   21            0     523.95            14.9     588.693625   
2         1.68   39            6     527.28            16.8     594.407400   
3         1.23   29            3    1056.18            12.3    1167.314400   
4        46.43   20            1    3902.00           464.3    4770.182750   

   line_cost  
0   80159.55  
1     331.79  
2     323.34  
3     732.95  
4    2627.70  

[5 rows x 22 columns]

In [177]:
# Create a column for the profit of each line using line cost and line price
df['line_profit'] = df['total_price'] - df['line_cost'] # calculate the profit for each line
display(df[['total_price', 'line_cost', 'line_profit']].head()) # this will print the first few rows to verify the new column



total_price  line_cost   line_profit
0  125898.607500   80159.55  45739.057500
1     588.693625     331.79    256.903625
2     594.407400     323.34    271.067400
3    1167.314400     732.95    434.364400
4    4770.182750    2627.70   2142.482750

## Part 3: Confirm your work
You have email receipts showing that the total prices for 3 orders. Confirm that your calculations match the receipts. Remember, each order has multiple lines.

Order ID 2742071 had a total price of \$152,811.89

Order ID 2173913 had a total price of \$162,388.71

Order ID 6128929 had a total price of \$923,441.25


In [180]:
# Check your work using the totals above

# Calculate the total price for each order
order_totals = df.groupby('order_id')['total_price'].sum()

# Check if the calculated totals match the given totals
order_2742071_total = order_totals.loc[2742071]
order_2173913_total = order_totals.loc[2173913]
order_6128929_total = order_totals.loc[6128929]

# Expected totals
expected_2742071_total = 152811.89
expected_2173913_total = 162388.71
expected_6128929_total = 923441.25

# Check for accuracy and calculate the difference
accuracy_2742071 = abs(order_2742071_total - expected_2742071_total)
accuracy_2173913 = abs(order_2173913_total - expected_2173913_total)
accuracy_6128929 = abs(order_6128929_total - expected_6128929_total)

# Print results
print(f"Order ID 2742071 calculated total: {order_2742071_total:.2f}")
print(f"Order ID 2742071 is {'accurate' if accuracy_2742071 < 1e-2 else 'an error'}, difference: {accuracy_2742071:.2f}")

print(f"Order ID 2173913 calculated total: {order_2173913_total:.2f}")
print(f"Order ID 2173913 is {'accurate' if accuracy_2173913 < 1e-2 else 'an error'}, difference: {accuracy_2173913:.2f}")

print(f"Order ID 6128929 calculated total: {order_6128929_total:.2f}")
print(f"Order ID 6128929 is {'accurate' if accuracy_6128929 < 1e-2 else 'an error'}, difference: {accuracy_6128929:.2f}")

# Expected outputs:
# Order ID 2742071 had a total price of $152,811.89
# Order ID 2173913 had a total price of $162,388.71
# Order ID 6128929 had a total price of $923,441.25


Order ID 2742071 calculated total: 139844.51
Order ID 2742071 is an error, difference: 12967.38
Order ID 2173913 calculated total: 119838.36
Order ID 2173913 is an error, difference: 42550.35
Order ID 6128929 calculated total: 752748.75
Order ID 6128929 is an error, difference: 170692.50


## Part 4: Summarize and Analyze
Use the new columns with confirmed values to find the following information.

In [183]:
# Calculate the total spending for each of the top 5 clients
client_spending = df[df['client_id'].isin(top_5_clients)].groupby('client_id')['total_price'].sum()

# Print the total spending for each of the top 5 clients
for client_id in top_5_clients:
    print(f"Client ID {client_id} spent: ${client_spending[client_id]:.2f}")



Client ID 33615 spent: $6392918.35
Client ID 66037 spent: $8746255.31
Client ID 46820 spent: $8006332.53
Client ID 38378 spent: $9172866.14
Client ID 24741 spent: $76678479.05


In [185]:
# Create a summary DataFrame showing the totals for the top 5 clients with the following information:
# total units purchased, total shipping price, total revenue, and total profit.

# Filter the DataFrame for the top 5 clients
top_5_clients_df = df[df['client_id'].isin(top_5_clients)]

# Group by client_id and calculate the required totals
summary_df = top_5_clients_df.groupby('client_id').agg({
    'qty': 'sum',             # total units purchased
    'shipping_price': 'sum',  # total shipping price
    'total_price': 'sum',     # total revenue
    'line_profit': 'sum'      # total profit
}).reset_index()

# Rename the columns for clarity
summary_df.columns = ['Client ID', 'Total Units Purchased', 'Total Shipping Price', 'Total Revenue', 'Total Profit']

# Convert Total Revenue and Total Profit to formatted string values with $ and commas
summary_df['Total Revenue'] = summary_df['Total Revenue'].apply(lambda x: f"${x:,.2f}")
summary_df['Total Profit'] = summary_df['Total Profit'].apply(lambda x: f"${x:,.2f}")

display(summary_df) # this will print the summary DataFrame




Client ID  Total Units Purchased  Total Shipping Price   Total Revenue  \
0      24741                 239862                9365.6  $76,678,479.05   
1      33615                  64313               12609.4   $6,392,918.35   
2      38378                  73667               11895.0   $9,172,866.14   
3      46820                  75768               11094.8   $8,006,332.53   
4      66037                  43018               10017.3   $8,746,255.31   

     Total Profit  
0  $36,106,662.11  
1   $2,033,979.93  
2   $2,955,705.56  
3   $2,589,494.93  
4   $3,126,906.88

In [187]:
# Format the data and rename the columns to names suitable for presentation.

# Define the money columns
money_columns = ['Total Revenue', 'Total Profit']

# Define a function that converts a dollar amount to millions
def currency_format_millions(value):
    return f"${value / 1_000_000:.2f}M"

# Apply the currency_format_millions function to only the money columns
summary_df[money_columns] = summary_df[money_columns].replace('[\$,]', '', regex=True).astype(float).applymap(currency_format_millions)

# Rename the columns to reflect the change in the money format
summary_df.columns = [
    'Client ID', 
    'Total Units Purchased', 
    'Total Shipping Price (USD)', 
    'Total Revenue (USD in Millions)', 
    'Total Profit (USD in Millions)'
]

display(summary_df) # this will print the formatted summary DataFrame





<>:11: SyntaxWarning: invalid escape sequence '\$'
<>:11: SyntaxWarning: invalid escape sequence '\$'
C:\Users\HP\AppData\Local\Temp\ipykernel_8384\4252629698.py:11: SyntaxWarning: invalid escape sequence '\$'
  summary_df[money_columns] = summary_df[money_columns].replace('[\$,]', '', regex=True).astype(float).applymap(currency_format_millions)
C:\Users\HP\AppData\Local\Temp\ipykernel_8384\4252629698.py:11: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  summary_df[money_columns] = summary_df[money_columns].replace('[\$,]', '', regex=True).astype(float).applymap(currency_format_millions)


Client ID  Total Units Purchased  Total Shipping Price (USD)  \
0      24741                 239862                      9365.6   
1      33615                  64313                     12609.4   
2      38378                  73667                     11895.0   
3      46820                  75768                     11094.8   
4      66037                  43018                     10017.3   

  Total Revenue (USD in Millions) Total Profit (USD in Millions)  
0                         $76.68M                        $36.11M  
1                          $6.39M                         $2.03M  
2                          $9.17M                         $2.96M  
3                          $8.01M                         $2.59M  
4                          $8.75M                         $3.13M

In [189]:
# Sort the updated data by "Total Profit (USD in Millions)" from highest to lowest and assign the sort to a new DataFrame
sorted_summary_df = summary_df.sort_values(by='Total Profit (USD in Millions)', ascending=False)

display(sorted_summary_df) # this will print the sorted summary DataFrame


Client ID  Total Units Purchased  Total Shipping Price (USD)  \
0      24741                 239862                      9365.6   
4      66037                  43018                     10017.3   
2      38378                  73667                     11895.0   
3      46820                  75768                     11094.8   
1      33615                  64313                     12609.4   

  Total Revenue (USD in Millions) Total Profit (USD in Millions)  
0                         $76.68M                        $36.11M  
4                          $8.75M                         $3.13M  
2                          $9.17M                         $2.96M  
3                          $8.01M                         $2.59M  
1                          $6.39M                         $2.03M